compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4020,14 +4009,17 @@ private BType checkArrayIndexBasedAccess(BLangIndexBasedAccess indexBasedAccess,
         BType actualType = symTable.semanticError;
         switch (indexExprType.tag) {
             case TypeTags.INT:
-                if (indexBasedAccess.indexExpr.getKind() != NodeKind.NUMERIC_LITERAL ||
-                        arrayType.state == BArrayState.UNSEALED) {
+                BLangExpression indexExpr = indexBasedAccess.indexExpr;
+                if (!isConst(indexExpr) || arrayType.state == BArrayState.UNSEALED) {
                     actualType = arrayType.eType;
                     break;
                 }
 
-                actualType = ((Long) ((BLangLiteral) indexBasedAccess.indexExpr).value) >= arrayType.size ?
-                        symTable.semanticError : arrayType.eType;
+                Long index = indexExpr.getKind() == NodeKind.NUMERIC_LITERAL ?
```

Comment:
I see a similar logic repeated in a few places, better to make it to a sperate set of functions.

Reply:
Fixed with 9333c81.